### PR TITLE
Add delegations

### PR DESCRIPTION
**Pull request by tanmay3011**

### CODE REVIEW

**Comment by cbrunsdon** [comment]
:+1:, thanks @tanmay3011! Really appreciate the effort trying to clean up core.

File: core/app/models/spree/stock_item.rb [hunk 1]
```diff
@@ -6,12 +6,15 @@ class StockItem < Spree::Base
     belongs_to :variant, class_name: 'Spree::Variant', inverse_of: :stock_items
     has_many :stock_movements, inverse_of: :stock_item
 
-    validates_presence_of :stock_location, :variant
-    validates_uniqueness_of :variant_id, scope: [:stock_location_id, :deleted_at]
+    validates :stock_location, :variant, presence: true
+    validates :variant_id, uniqueness: { scope: [:stock_location_id, :deleted_at] }, allow_blank: true
```
Comment (Senjai) [inline core/app/models/spree/stock_item.rb]: Did we want this `allow_blank: true` to be a part of the uniqueness clause?
`validates :variant_id, uniqueness: { scope: [:stock_location_id, :deleted_at], allow_blank: true }`

Comment (Senjai) [inline core/app/models/spree/stock_item.rb]: Ah, this is valid. allow_blank can be passed directly into the `UniquenessValidator` or used as an option to the `validates` method. The difference being whether or not the validators are invoked, or are expected to handle `:allow_blank` directly (which uniqueness does). I've only ever passed allow_blank into specific validators, I didn't know you could short circuit using it outside of those validation statements. :+1:

**Comment by Senjai** [comment]
:shipit: Good work! @tanmay3011

**Comment by athal7** [comment]
:+1:. Was a bit thrown by the allow_blank and allow_nil paired with the presence validation. The performance benefits make sense to me, I wonder if a comment would help to not trip up future contributors though?

**Comment by tanmay3011** [comment]
@athal7 Where should I add comment to each file where I have used allow_blank or to any specific place

**Comment by athal7** [comment]
@tanmay3011 hm yea that does seem excessive. would it save the extra query if you removed the presence validation instead and changed allow_blank to false? that would be a bit less confusing IMO.

If not, :+1:

**Comment by tanmay3011** [comment]
@athal7 presence validation shouldn't be removed since case of nil won't be handled in uniqueness validation. So presence validation is a must.

**Comment by athal7** [comment]
@tanmay3011 I think I'm missing where the query count is reduced, can you say more? Here's my own little test:

before adding allow_blank:
```ruby
irb(main):001:0> Spree::AdjustmentReason.create!(name: "foo", code: "bar")
   (0.0ms)  begin transaction
  Spree::AdjustmentReason Exists (0.1ms)  SELECT  1 AS one FROM "spree_adjustment_reasons" WHERE LOWER("spree_adjustment_reasons"."name") = LOWER('foo') LIMIT 1
  Spree::AdjustmentReason Exists (0.1ms)  SELECT  1 AS one FROM "spree_adjustment_reasons" WHERE LOWER("spree_adjustment_reasons"."code") = LOWER('bar') LIMIT 1
  SQL (0.3ms)  INSERT INTO "spree_adjustment_reasons" ("name", "code", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["name", "foo"], ["code", "bar"], ["created_at", "2015-12-07 14:26:44.287782"], ["updated_at", "2015-12-07 14:26:44.287782"]]
   (0.8ms)  commit transaction
```

after adding allow_blank:
```ruby
irb(main):002:0> Spree::AdjustmentReason.create!(name: "foo", code: "bar")
   (0.1ms)  begin transaction
  Spree::AdjustmentReason Exists (0.1ms)  SELECT  1 AS one FROM "spree_adjustment_reasons" WHERE LOWER("spree_adjustment_reasons"."name") = LOWER('foo') LIMIT 1
  Spree::AdjustmentReason Exists (0.1ms)  SELECT  1 AS one FROM "spree_adjustment_reasons" WHERE LOWER("spree_adjustment_reasons"."code") = LOWER('bar') LIMIT 1
  SQL (0.4ms)  INSERT INTO "spree_adjustment_reasons" ("name", "code", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["name", "foo"], ["code", "bar"], ["created_at", "2015-12-07 14:20:24.484747"], ["updated_at", "2015-12-07 14:20:24.484747"]]
   (1.5ms)  commit transaction
```

**Comment by tanmay3011** [comment]
@athal7 Did you reload your console? 
Inorder to reload all your models again with the validation in affect

**Comment by athal7** [comment]
I did, and also entirely rebuilt the sandbox just to make double sure.

**Comment by tanmay3011** [comment]
@athal7  Just see the example I posted. You need to leave name and code attribute blank in both cases to see the difference

**Comment by athal7** [comment]
@tanmay3011 thanks for the example, that helps / makes sense. 

Since we are trying to prevent blank names and codes, my personal preference would be to introduce a client-side validation and avoid db interaction entirely rather than introducing this level of indirection. I know that doesn't cover us in the API, but personally I prefer that trade-off. 

Curious to hear opinions of other core team members...

**Comment by cbrunsdon** [comment]
@athal7 I've got nothing against it. Its definitely a quirk/optimization of rails I hadn't considered before but am for it. @gmacdougall @jordan-brough @jhawthorn ?

**Comment by gmacdougall** [comment]
I'm not particularly bothered about the extra validation queries either way. I do like having validations live as close to the DB as possible, as front end validations can easily be bypassed.

**Comment by athal7** [comment]
@gmacdougall just to be clear, I'm not suggesting we remove any sort of validating code, just remove seemingly contradictory validations that are in place to reduce queries in cases that we don't expect. the presence: true would stay.

**Comment by gmacdougall** [comment]
Thanks @athal7  I wasn't clear on that as I was having trouble following the flow of the conversation here.

**Comment by tanmay3011** [comment]
@athal7 Do I need to do any changes now?

**Comment by athal7** [comment]
I would be in favor of the changes I recommended earlier, but not going to
stand in front of this if the rest of the core team feels strongly
otherwise.

**Comment by jhawthorn** [comment]
I am in favour of the changes here as they are now. Though the `allow_blank` might read as contradictory, it does behave exactly as one would expect, ie. not checking checking uniqueness on an already obviously invalid record.

**Comment by tanmay3011** [comment]
@jhawthorn Ok! Will update the PR and remove conflicts

**Comment by jhawthorn** [comment]
Merged as 2ad85706d129f4e67a7dd568a786f4aa2326fc29. Thanks!

**Pull request closed without merging**